File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/02_trees/ValidateBst.cs
// LeetCode 98 - Validate Binary Search Tree
// Difficulty: Medium
// Pattern: DFS with min/max bounds
//
// Approach: Pass valid range [minVal, maxVal] down the recursion.
// Each node must be strictly between its bounds.
//
// Time: O(n)  Space: O(h)

namespace CodingPatterns.Trees;

public class TreeNode
{
    public int Val;
    public TreeNode Left, Right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    { Val = val; Left = left; Right = right; }
}

public class ValidateBST
{
    public bool IsValidBST(TreeNode root)
        => Validate(root, long.MinValue, long.MaxValue);

    private bool Validate(TreeNode node, long minVal, long maxVal)
    {
        if (node == null) return true;
        if (node.Val <= minVal || node.Val >= maxVal) return false;
        return Validate(node.Left, minVal, node.Val) &&
               Validate(node.Right, node.Val, maxVal);
    }

    public bool IsValidBSTOptimized(TreeNode root)
    {
        var stack = new Stack<TreeNode>();
        var current = root;
        long prev = long.MinValue;

        while (current != null || stack.Count > 0)
        {
            // Walk down all the way left
            while (current != null)
            {
                stack.Push(current);
                current = current.Left;
            }

            // Visit the smallest unvisited node
            current = stack.Pop();
            if (current.Val <= prev) return false;
            prev = current.Val;

            // Move to right subtree
            current = current.Right;
        }
        return true;
    }

    public static void Run()
    {
        var sol = new ValidateBST();
        var valid = new TreeNode(2, new TreeNode(1), new TreeNode(3));
        Console.WriteLine(sol.IsValidBST(valid)); // True

        var invalid = new TreeNode(5, new TreeNode(1),
            new TreeNode(4, new TreeNode(3), new TreeNode(6)));
        Console.WriteLine(sol.IsValidBST(invalid)); // False
    }
}

File: Coding-Patterns/Coding-Patterns-Python/02_trees/validate_bst.py
# LeetCode 98 - Validate Binary Search Tree
# Difficulty: Medium
# Pattern: DFS with min/max bounds
#
# Problem: Given the root of a binary tree, determine if it is a valid BST.
#
# Approach: Pass down valid range [min_val, max_val] at each node.
# Each node must satisfy min_val < node.val < max_val.
# Left child tightens max, right child tightens min.
#
# Time: O(n)  Space: O(h) where h = tree height

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_valid_bst(root: Optional[TreeNode]) -> bool:
    def validate(node, min_val, max_val):
        if not node:
            return True
        if not (min_val < node.val < max_val):
            return False
        return (validate(node.left, min_val, node.val) and
                validate(node.right, node.val, max_val))

    return validate(root, float('-inf'), float('inf'))


# ---- Tests ----
if __name__ == "__main__":
    # Valid BST:   2
    #            / \
    #           1   3
    root = TreeNode(2, TreeNode(1), TreeNode(3))
    assert is_valid_bst(root) == True

    # Invalid:   5
    #           / \
    #          1   4
    #             / \
    #            3   6
    root2 = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
    assert is_valid_bst(root2) == False
    print("All tests passed.")
